[Security Solution][Detections] Rule forms cleanup

rylnd opened the pull request:
## Summary

In preparation for new rule types/fields, and in response to improvements to the form library, this adds types to the majority of our form code.

 ~~Besides the bugfixes,the main behavioral change here is that it is no longer possible to navigate between invalid tabs on the Rule Edit page, and you must instead make a tab valid before moving away. This is how the Create page currently behaves, and so behavior (and implementation) are now consolidated between the two pages.~~ **edit: I've reverted this behavior but changed the invalid callout to display on navigation between tabs (previously it was only shown when you clicked Save.** Once we have UX for displaying validation errors to the user on both Create and Edit, we can persist invalid data in a consistent manner.

### "Features": 
* adds types our form state/helpers
  * (we had a lot of implicit `any`s and `as`s in there)
* adds regression assertions to an existing "create rule" cypress test (7fa8bca)
* Removes intermediate state in our form components
  * Summary: the parent holds the data to be persisted, while the form lib holds the data as the user interacts with the form; most of the local state was there to work around issues with the form lib (e.g. memoization, which is now fixed)
  * We persist form hooks (`RuleStepsFormHooks`)  in the parent so that we can validate and store the active step's data
  * We persist VALID form data (`RuleStepsFormData`) in the parent in order to repopulate a form that has previously unmounted
    * If a form is unmounted _or a form's field is unmounted_, the form hook loses knowledge of the form/field
* Removes uses of FormDataProvider
  * These were used to react to form changes, and typically used local state to track the old value to compare.
  * In (almost) all of our current uses, we only care about current and default values, never the "previous value," and so we were able to reimplement the same functionality with `useFormData` and `useEffect`.

### "Bugfixes":
* Fixes moving away from the Actions step (https://github.com/elastic/kibana/pull/76138/commits/b1742706f839c91dffc6b9a7811f080ecc1b7158)
* Fixes bug moving between steps of the Edit form (https://github.com/elastic/kibana/pull/76138/commits/4995e814c7e8b7169bbdc398f0f2bade327339d0)
  * Consequence here is that if a user navigates away from an invalid step, their modifications will be lost
* Fixes "link" between default severity and default risk score, where changing the former updated the latter (https://github.com/elastic/kibana/pull/76138/commits/0972e16230906e475a208817d8cfa4a709259600)
* Removes some unnecessary ML API calls (aac72c939dc57974468195515d47a4337e427dd8)


### Unaddressed issues/notes
- [ ] With the severity/risk score link back in place, there's now an interesting UI oddity: the default severity is `low`, but the default risk score is `50`. However, the default _mapped_ risk score for `low` is `21`. This might be fine, I just wanted to bring it up as a potential discrepancy.
- [x] I believe that there may be a validation bug in the form library.  **Update**: fixed with https://github.com/elastic/kibana/pull/76379
    1. Have a text field with an `emptyField` validation (e.g. the Rule Name field)
    1. leave the field blank, submit the form, observe the validation error
    1. enter text to remove the validation error
    1. delete text
    1. submit the form and see that a `""` has been persisted to that field
    Update: should be fixed with https://github.com/elastic/kibana/pull/76379

### Checklist

- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)


### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71001/)
* Commit: ad8aa6059fc5f8422bfe487a03f1fbee77ede92a

### Build metrics
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/ad8aa6059fc5f8422bfe487a03f1fbee77ede92a) | diff | [baseline](https://github.com/elastic/kibana/commit/3541e779c62a9f4f0d398a7e41316cc1f33018bb) |
| --- | --- | --- | --- |
| securitySolution | 1926 | +1 | 1925 |

#### async chunks size
| id | [value](https://github.com/elastic/kibana/commit/ad8aa6059fc5f8422bfe487a03f1fbee77ede92a) | diff | [baseline](https://github.com/elastic/kibana/commit/3541e779c62a9f4f0d398a7e41316cc1f33018bb) |
| --- | --- | --- | --- |
| securitySolution | 9.5MB | -6.4KB | 9.5MB |

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71001/","number":"71001","commit":"ad8aa6059fc5f8422bfe487a03f1fbee77ede92a"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71086/)
* Commit: 7fa8bca430263e0d379ade0ab60cb738c40e800f

### Build metrics
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/7fa8bca430263e0d379ade0ab60cb738c40e800f) | diff | [baseline](https://github.com/elastic/kibana/commit/3541e779c62a9f4f0d398a7e41316cc1f33018bb) |
| --- | --- | --- | --- |
| securitySolution | 1926 | +1 | 1925 |

#### async chunks size
| id | [value](https://github.com/elastic/kibana/commit/7fa8bca430263e0d379ade0ab60cb738c40e800f) | diff | [baseline](https://github.com/elastic/kibana/commit/3541e779c62a9f4f0d398a7e41316cc1f33018bb) |
| --- | --- | --- | --- |
| securitySolution | 9.5MB | -6.2KB | 9.5MB |

### History
* :green_heart: [Build #71001](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71001/) succeeded ad8aa6059fc5f8422bfe487a03f1fbee77ede92a

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71001/","number":"71001","commit":"ad8aa6059fc5f8422bfe487a03f1fbee77ede92a"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71086/","number":"71086","commit":"7fa8bca430263e0d379ade0ab60cb738c40e800f"}]}
PIPELINE-->

rylnd commented:
@patrykkopycinski @sebelga I'm seeing some odd behavior, described in the first "Unaddressed issues/notes" above. I haven't yet determined whether it's a bug with the form lib or merely with our forms, but it's present on both this branch and master. If either of you has a moment to :eyes: that issue and provide some guidance it would be much appreciated.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71474/)
* Commit: 453ef00b32c097fa8e926b347dcb04a561ba832b

### Build metrics
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/453ef00b32c097fa8e926b347dcb04a561ba832b) | diff | [baseline](https://github.com/elastic/kibana/commit/3541e779c62a9f4f0d398a7e41316cc1f33018bb) |
| --- | --- | --- | --- |
| securitySolution | 1927 | +2 | 1925 |

#### async chunks size
| id | [value](https://github.com/elastic/kibana/commit/453ef00b32c097fa8e926b347dcb04a561ba832b) | diff | [baseline](https://github.com/elastic/kibana/commit/3541e779c62a9f4f0d398a7e41316cc1f33018bb) |
| --- | --- | --- | --- |
| securitySolution | 9.5MB | -10.4KB | 9.5MB |

### History
* :green_heart: [Build #71086](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71086/) succeeded 7fa8bca430263e0d379ade0ab60cb738c40e800f
* :green_heart: [Build #71001](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71001/) succeeded ad8aa6059fc5f8422bfe487a03f1fbee77ede92a

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71001/","number":"71001","commit":"ad8aa6059fc5f8422bfe487a03f1fbee77ede92a"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71086/","number":"71086","commit":"7fa8bca430263e0d379ade0ab60cb738c40e800f"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71474/","number":"71474","commit":"453ef00b32c097fa8e926b347dcb04a561ba832b"}]}
PIPELINE-->

sebelga commented:
@rylnd Having a look...

sebelga commented on x-pack/plugins/security_solution/public/detections/components/rules/step_rule_actions/index.tsx:
```diff
@@ -94,18 +93,21 @@ const StepRuleActionsComponent: FC<StepRuleActionsProps> = ({
-          setStepData(RuleStep.ruleActions, { ...data, enabled }, newIsValid);
-          setMyStepData({ ...data, isNew: false } as ActionsStepRule);
-        }
+  useEffect(() => {
```
This `useEffect` looks strange. The field should get its value from `defaultValue` defined on the form.

rylnd replied:
Good call, this was moved to `defaultValue`.

sebelga commented on x-pack/plugins/security_solution/public/detections/components/rules/step_rule_actions/index.tsx:
```diff
@@ -114,16 +116,18 @@ const StepRuleActionsComponent: FC<StepRuleActionsProps> = ({
-    myStepData,
-    setMyStepData,
-  ]);
+  const updateThrottle = useCallback(
```
We might need to zoom as we never use `getFields()` in our forms, and I see that you do in need it in multiple places and manually set the value `setValue()`

rylnd replied:
Yep, this one was not needed (https://github.com/elastic/kibana/pull/76138/commits/f36baf564812b2377e8cc453c546b49661688f8f), and the local state was later replaced with `useFormData` 👍

sebelga commented:
@rylnd I saw what it is. I introduced a regression in my last PR and will revert it asap! Sorry for this and thanks for raising it 👍

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71757/)
* Commit: 3d64b012f476ab472cea79d42b6a4962bb2f664f
* This comment will update when the build is complete

### Failed CI Steps
* [Execute xpack-securitySolutionCypress](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71757/execution/node/589/log)
* [Execute xpack-securitySolutionCypress](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71757/execution/node/747/log)

### History
* :green_heart: [Build #71474](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71474/) succeeded 453ef00b32c097fa8e926b347dcb04a561ba832b
* :green_heart: [Build #71086](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71086/) succeeded 7fa8bca430263e0d379ade0ab60cb738c40e800f
* :green_heart: [Build #71001](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71001/) succeeded ad8aa6059fc5f8422bfe487a03f1fbee77ede92a

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71001/","number":"71001","commit":"ad8aa6059fc5f8422bfe487a03f1fbee77ede92a"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71086/","number":"71086","commit":"7fa8bca430263e0d379ade0ab60cb738c40e800f"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71474/","number":"71474","commit":"453ef00b32c097fa8e926b347dcb04a561ba832b"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71757/","number":"71757","commit":"3d64b012f476ab472cea79d42b6a4962bb2f664f"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72121/)
* Commit: 36ba03fa1330c7bd3dd5c0fe129d8a503f07a931

### Build metrics
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/36ba03fa1330c7bd3dd5c0fe129d8a503f07a931) | diff | [baseline](https://github.com/elastic/kibana/commit/9a27beba75a57e39caeb3dcfa1c57000da6cb630) |
| --- | --- | --- | --- |
| securitySolution | 1934 | +1 | 1933 |

#### async chunks size
| id | [value](https://github.com/elastic/kibana/commit/36ba03fa1330c7bd3dd5c0fe129d8a503f07a931) | diff | [baseline](https://github.com/elastic/kibana/commit/9a27beba75a57e39caeb3dcfa1c57000da6cb630) |
| --- | --- | --- | --- |
| securitySolution | 9.5MB | -8.3KB | 9.5MB |

### History
* :broken_heart: [Build #71757](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71757/) failed 3d64b012f476ab472cea79d42b6a4962bb2f664f
* :green_heart: [Build #71474](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71474/) succeeded 453ef00b32c097fa8e926b347dcb04a561ba832b
* :green_heart: [Build #71086](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71086/) succeeded 7fa8bca430263e0d379ade0ab60cb738c40e800f
* :green_heart: [Build #71001](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71001/) succeeded ad8aa6059fc5f8422bfe487a03f1fbee77ede92a

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71001/","number":"71001","commit":"ad8aa6059fc5f8422bfe487a03f1fbee77ede92a"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71086/","number":"71086","commit":"7fa8bca430263e0d379ade0ab60cb738c40e800f"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71474/","number":"71474","commit":"453ef00b32c097fa8e926b347dcb04a561ba832b"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71757/","number":"71757","commit":"3d64b012f476ab472cea79d42b6a4962bb2f664f"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72121/","number":"72121","commit":"36ba03fa1330c7bd3dd5c0fe129d8a503f07a931"}]}
PIPELINE-->

elasticmachine commented:
Pinging @elastic/siem (Team:SIEM)

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72355/)
* Commit: 33cdeef0562efd21813209d9b3ffef64c46e8fb8

### Build metrics
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/33cdeef0562efd21813209d9b3ffef64c46e8fb8) | diff | [baseline](https://github.com/elastic/kibana/commit/8b085b9eac95d3ebeedeee3b3963c7876a9e0e12) |
| --- | --- | --- | --- |
| securitySolution | 1942 | +1 | 1941 |

#### async chunks size
| id | [value](https://github.com/elastic/kibana/commit/33cdeef0562efd21813209d9b3ffef64c46e8fb8) | diff | [baseline](https://github.com/elastic/kibana/commit/8b085b9eac95d3ebeedeee3b3963c7876a9e0e12) |
| --- | --- | --- | --- |
| securitySolution | 9.8MB | -7.7KB | 9.8MB |

### History
* :green_heart: [Build #72121](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72121/) succeeded 36ba03fa1330c7bd3dd5c0fe129d8a503f07a931
* :broken_heart: [Build #71757](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71757/) failed 3d64b012f476ab472cea79d42b6a4962bb2f664f
* :green_heart: [Build #71474](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71474/) succeeded 453ef00b32c097fa8e926b347dcb04a561ba832b
* :green_heart: [Build #71086](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71086/) succeeded 7fa8bca430263e0d379ade0ab60cb738c40e800f
* :green_heart: [Build #71001](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71001/) succeeded ad8aa6059fc5f8422bfe487a03f1fbee77ede92a

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71001/","number":"71001","commit":"ad8aa6059fc5f8422bfe487a03f1fbee77ede92a"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71086/","number":"71086","commit":"7fa8bca430263e0d379ade0ab60cb738c40e800f"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71474/","number":"71474","commit":"453ef00b32c097fa8e926b347dcb04a561ba832b"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71757/","number":"71757","commit":"3d64b012f476ab472cea79d42b6a4962bb2f664f"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72121/","number":"72121","commit":"36ba03fa1330c7bd3dd5c0fe129d8a503f07a931"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72355/","number":"72355","commit":"33cdeef0562efd21813209d9b3ffef64c46e8fb8"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72404/)
* Commit: db36f407c6b576001df6c1f95ccd00c415c6a964

### Build metrics
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/db36f407c6b576001df6c1f95ccd00c415c6a964) | diff | [baseline](https://github.com/elastic/kibana/commit/8b085b9eac95d3ebeedeee3b3963c7876a9e0e12) |
| --- | --- | --- | --- |
| securitySolution | 1942 | +1 | 1941 |

#### async chunks size
| id | [value](https://github.com/elastic/kibana/commit/db36f407c6b576001df6c1f95ccd00c415c6a964) | diff | [baseline](https://github.com/elastic/kibana/commit/8b085b9eac95d3ebeedeee3b3963c7876a9e0e12) |
| --- | --- | --- | --- |
| securitySolution | 9.8MB | +1.7KB | 9.8MB |

### History
* :green_heart: [Build #72355](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72355/) succeeded 33cdeef0562efd21813209d9b3ffef64c46e8fb8
* :green_heart: [Build #72121](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72121/) succeeded 36ba03fa1330c7bd3dd5c0fe129d8a503f07a931
* :broken_heart: [Build #71757](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71757/) failed 3d64b012f476ab472cea79d42b6a4962bb2f664f
* :green_heart: [Build #71474](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71474/) succeeded 453ef00b32c097fa8e926b347dcb04a561ba832b
* :green_heart: [Build #71086](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71086/) succeeded 7fa8bca430263e0d379ade0ab60cb738c40e800f

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71086/","number":"71086","commit":"7fa8bca430263e0d379ade0ab60cb738c40e800f"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71474/","number":"71474","commit":"453ef00b32c097fa8e926b347dcb04a561ba832b"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71757/","number":"71757","commit":"3d64b012f476ab472cea79d42b6a4962bb2f664f"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72121/","number":"72121","commit":"36ba03fa1330c7bd3dd5c0fe129d8a503f07a931"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72355/","number":"72355","commit":"33cdeef0562efd21813209d9b3ffef64c46e8fb8"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72404/","number":"72404","commit":"db36f407c6b576001df6c1f95ccd00c415c6a964"}]}
PIPELINE-->

spong commented:
No need to address here (so much goodness already! :), but wanted to comment on the standardization of the disabled/loading state of fields. Looks like we're closer than before, but there are still a few fields that don't disable when loading/saving. This may be particularly hard to get the `Actions` component to match since we're pulling in the alerting component full sail, but we should be able to get the others to match since we own those components. In the screenshots below, the arrows point to the fields that don't disable on load/save:

<details><summary>Definition</summary>
<p>

<p align="center">
  <img width="500" src="https://user-images.githubusercontent.com/2946766/92269172-e86f2e80-eea0-11ea-8875-a257d77d011a.png" />
</p>
</p>
</details>


<details><summary>About</summary>
<p>

<p align="center">
  <img width="500" src="https://user-images.githubusercontent.com/2946766/92268795-451e1980-eea0-11ea-97db-d4872da5412b.png" />
</p>
</p>
</details>

<details><summary>Schedule (As above, probably not much we can do here unless the alerting component supports the loading/disabled state)</summary>
<p>

<p align="center">
  <img width="500" src="https://user-images.githubusercontent.com/2946766/92268635-fbcdca00-ee9f-11ea-9417-6664f8c65986.png" />
</p>
</p>
</details>

<details><summary>Actions</summary>
<p>

<p align="center">
  <img width="500" src="https://user-images.githubusercontent.com/2946766/92268670-0daf6d00-eea0-11ea-9dbb-1bdd0556c3be.png" />
</p>
</p>
</details>

spong commented on x-pack/plugins/security_solution/public/detections/components/rules/next_step/index.tsx:
```diff
@@ -9,7 +9,7 @@ import { EuiHorizontalRule, EuiFlexGroup, EuiFlexItem, EuiButton } from '@elasti
 
 interface NextStepProps {
-  onClick: () => Promise<void>;
+  onClick: () => void;
```
Nice catch!

spong commented on x-pack/plugins/security_solution/public/detections/components/rules/step_about_rule/data.tsx:
```diff
@@ -8,10 +8,9 @@ import styled from 'styled-components';
 import { EuiHealth } from '@elastic/eui';
 import euiLightVars from '@elastic/eui/dist/eui_theme_light.json';
 import React from 'react';
+import { SeverityValue } from '../../../pages/detection_engine/rules/types';
```
Would it be possible to start using the `common` types here?

e.g. 

https://github.com/spong/kibana/blob/d2156b7285c0397fe93a12cd8f1464ac430bc0e8/x-pack/plugins/security_solution/common/detection_engine/schemas/common/schemas.ts#L256-L257

spong commented on x-pack/plugins/security_solution/public/detections/components/rules/step_about_rule/index.tsx:
```diff
@@ -68,47 +69,69 @@ const StepAboutRuleComponent: FC<StepAboutRuleProps> = ({
-        setStepData(RuleStep.aboutRule, data, isValid);
-        setMyStepData({ ...data, isNew: false } as AboutStepRule);
+  useEffect(() => {
+    const formSeverityValue = formSeverity?.value;
+    if (formSeverityValue != null && formSeverityValue !== severityValue) {
+      setSeverityValue(formSeverityValue);
+
+      const newRiskScoreValue = defaultRiskScoreBySeverity[formSeverityValue];
+      if (newRiskScoreValue != null) {
+        const riskScoreField = getFields().riskScore as FieldHook<AboutStepRule['riskScore']>;
+        riskScoreField.setValue({ ...riskScoreField.value, value: newRiskScoreValue });
       }
     }
-  }, [setStepData, submit]);
+  }, [formSeverity?.value, getFields, severityValue]);
```
Thanks for snagging this one and fixing the linking of the default severity+risk_score. As for your callout from the description:

> With the severity/risk score link back in place, there's now an interesting UI oddity: the default severity is low, but the default risk score is 50. However, the default mapped risk score for low is 21. This might be fine, I just wanted to bring it up as a potential discrepancy.

I think we'll want to  go ahead and set the default `risk_score` to match the linked default severity, so have it set to `21`. This will ensure users that select other default severities (and don't modify the default risk_score) will keep consistent with the linked values.

rylnd replied:
Done! e45c128

spong commented on x-pack/plugins/security_solution/public/detections/components/rules/step_rule_actions/index.tsx:
```diff
@@ -65,84 +64,86 @@ const StepRuleActionsComponent: FC<StepRuleActionsProps> = ({
+    schema,
+  });
+  const { getFields, getFormData, submit } = form;
+  const [{ throttle: formThrottle }] = (useFormData({
+    form,
+    watch: ['throttle'],
+  }) as unknown) as [Partial<ActionsStepRule>];
```
These always get a double-take from me when I see them, but understand with the removal of `FormDataProvider`.

rylnd replied:
Yeah, I've discussed this with @sebelga and I'm hopeful that we can get a generic type for our use case (where form data `F` is identical to output data `T`.

spong commented on x-pack/plugins/security_solution/public/detections/pages/detection_engine/rules/helpers.tsx:
```diff
@@ -154,7 +151,7 @@ export const getAboutStepsData = (rule: Rule, detailsView: boolean): AboutStepRu
     references,
     severity: {
-      value: severity,
+      value: severity as SeverityValue,
```
As above, would it be possible to start using the common types here?

e.g.

https://github.com/spong/kibana/blob/d2156b7285c0397fe93a12cd8f1464ac430bc0e8/x-pack/plugins/security_solution/common/detection_engine/schemas/common/schemas.ts#L256-L257

spong commented on x-pack/plugins/security_solution/public/detections/pages/detection_engine/rules/types.ts:
```diff
@@ -87,8 +103,10 @@ export interface AboutStepRuleDetails {
+
 export interface AboutStepSeverity {
-  value: string;
+  value: SeverityValue;
```
Last one on the `SeverityValue` promise... 😅 , but would it be possible to start using the `common` types here?

e.g.

https://github.com/spong/kibana/blob/d2156b7285c0397fe93a12cd8f1464ac430bc0e8/x-pack/plugins/security_solution/common/detection_engine/schemas/common/schemas.ts#L256-L257

spong reviewed (approved):
Checked out, tested locally, and performed code review. So much goodness here @rylnd, fantastic work with the cleanup effort! 🙂 Transforming that tech debt into a tech investment with all the type and control flow enhancements you packed in here. I've left a few nits (mostly about starting to use our `common` types if we can), but nothing worth holding this PR up. LGTM! 👍 🚀 🎉

rylnd commented:
Good catch on the SeverityValue redundancy. Fixed in https://github.com/elastic/kibana/pull/76138/commits/b24742cd7dc489add2ad7190d0843a67bc19675f.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72718/)
* Commit: 7d359ed8d90f23878c27e7583c07541ff588998d

### Build metrics
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/7d359ed8d90f23878c27e7583c07541ff588998d) | diff | [baseline](https://github.com/elastic/kibana/commit/6e0b20cbe3a2a0645cec4421265e165b76afe2df) |
| --- | --- | --- | --- |
| securitySolution | 1951 | +1 | 1950 |

#### async chunks size
| id | [value](https://github.com/elastic/kibana/commit/7d359ed8d90f23878c27e7583c07541ff588998d) | diff | [baseline](https://github.com/elastic/kibana/commit/6e0b20cbe3a2a0645cec4421265e165b76afe2df) |
| --- | --- | --- | --- |
| securitySolution | 9.9MB | +1.7KB | 9.9MB |

### History
* :green_heart: [Build #72404](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72404/) succeeded db36f407c6b576001df6c1f95ccd00c415c6a964
* :green_heart: [Build #72355](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72355/) succeeded 33cdeef0562efd21813209d9b3ffef64c46e8fb8
* :green_heart: [Build #72121](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72121/) succeeded 36ba03fa1330c7bd3dd5c0fe129d8a503f07a931
* :broken_heart: [Build #71757](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71757/) failed 3d64b012f476ab472cea79d42b6a4962bb2f664f
* :green_heart: [Build #71474](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71474/) succeeded 453ef00b32c097fa8e926b347dcb04a561ba832b

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71474/","number":"71474","commit":"453ef00b32c097fa8e926b347dcb04a561ba832b"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71757/","number":"71757","commit":"3d64b012f476ab472cea79d42b6a4962bb2f664f"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72121/","number":"72121","commit":"36ba03fa1330c7bd3dd5c0fe129d8a503f07a931"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72355/","number":"72355","commit":"33cdeef0562efd21813209d9b3ffef64c46e8fb8"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72404/","number":"72404","commit":"db36f407c6b576001df6c1f95ccd00c415c6a964"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72718/","number":"72718","commit":"7d359ed8d90f23878c27e7583c07541ff588998d"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72740/)
* Commit: e45c1288db2708b27d18b4de0343e3e2549a5b2e

### Build metrics
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/e45c1288db2708b27d18b4de0343e3e2549a5b2e) | diff | [baseline](https://github.com/elastic/kibana/commit/6e0b20cbe3a2a0645cec4421265e165b76afe2df) |
| --- | --- | --- | --- |
| securitySolution | 1951 | +1 | 1950 |

#### async chunks size
| id | [value](https://github.com/elastic/kibana/commit/e45c1288db2708b27d18b4de0343e3e2549a5b2e) | diff | [baseline](https://github.com/elastic/kibana/commit/6e0b20cbe3a2a0645cec4421265e165b76afe2df) |
| --- | --- | --- | --- |
| securitySolution | 9.9MB | +1.7KB | 9.9MB |

### History
* :green_heart: [Build #72718](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72718/) succeeded 7d359ed8d90f23878c27e7583c07541ff588998d
* :green_heart: [Build #72404](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72404/) succeeded db36f407c6b576001df6c1f95ccd00c415c6a964
* :green_heart: [Build #72355](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72355/) succeeded 33cdeef0562efd21813209d9b3ffef64c46e8fb8
* :green_heart: [Build #72121](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72121/) succeeded 36ba03fa1330c7bd3dd5c0fe129d8a503f07a931
* :broken_heart: [Build #71757](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71757/) failed 3d64b012f476ab472cea79d42b6a4962bb2f664f

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71757/","number":"71757","commit":"3d64b012f476ab472cea79d42b6a4962bb2f664f"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72121/","number":"72121","commit":"36ba03fa1330c7bd3dd5c0fe129d8a503f07a931"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72355/","number":"72355","commit":"33cdeef0562efd21813209d9b3ffef64c46e8fb8"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72404/","number":"72404","commit":"db36f407c6b576001df6c1f95ccd00c415c6a964"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72718/","number":"72718","commit":"7d359ed8d90f23878c27e7583c07541ff588998d"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72740/","number":"72740","commit":"e45c1288db2708b27d18b4de0343e3e2549a5b2e"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72745/)
* Commit: b24742cd7dc489add2ad7190d0843a67bc19675f

### Build metrics
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/b24742cd7dc489add2ad7190d0843a67bc19675f) | diff | [baseline](https://github.com/elastic/kibana/commit/6e0b20cbe3a2a0645cec4421265e165b76afe2df) |
| --- | --- | --- | --- |
| securitySolution | 1951 | +1 | 1950 |

#### async chunks size
| id | [value](https://github.com/elastic/kibana/commit/b24742cd7dc489add2ad7190d0843a67bc19675f) | diff | [baseline](https://github.com/elastic/kibana/commit/6e0b20cbe3a2a0645cec4421265e165b76afe2df) |
| --- | --- | --- | --- |
| securitySolution | 9.9MB | +1.7KB | 9.9MB |

### History
* :green_heart: [Build #72740](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72740/) succeeded e45c1288db2708b27d18b4de0343e3e2549a5b2e
* :green_heart: [Build #72718](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72718/) succeeded 7d359ed8d90f23878c27e7583c07541ff588998d
* :green_heart: [Build #72404](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72404/) succeeded db36f407c6b576001df6c1f95ccd00c415c6a964
* :green_heart: [Build #72355](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72355/) succeeded 33cdeef0562efd21813209d9b3ffef64c46e8fb8
* :green_heart: [Build #72121](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72121/) succeeded 36ba03fa1330c7bd3dd5c0fe129d8a503f07a931

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72121/","number":"72121","commit":"36ba03fa1330c7bd3dd5c0fe129d8a503f07a931"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72355/","number":"72355","commit":"33cdeef0562efd21813209d9b3ffef64c46e8fb8"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72404/","number":"72404","commit":"db36f407c6b576001df6c1f95ccd00c415c6a964"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72718/","number":"72718","commit":"7d359ed8d90f23878c27e7583c07541ff588998d"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72740/","number":"72740","commit":"e45c1288db2708b27d18b4de0343e3e2549a5b2e"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72745/","number":"72745","commit":"b24742cd7dc489add2ad7190d0843a67bc19675f"}]}
PIPELINE-->

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)
